### PR TITLE
[CIS-1878] Automate UI Scenarios

### PR DESCRIPTION
**Pull request by bielikb**
### 🔗 Issue Links

CIS-1878

### 🎯 Goal

Automate UI scenarios.

### Following UI scenarios were automated:

* User can send message with multiple lines
* Composer does not grow beyond 5 lines
* Typing indicator is shown when user starts typing
* Typing indicator is hidden when user stops typing
* Message list is scrolled down when user sends a new message
* Message list is scrolled down when participant sends a new message
* Message list does not scroll down when message list is scrolled up and user receives new message
* Quoted reply is shown in list when user adds a quote reply to participant's message
* Quoted reply is deleted when deleted by participant
* Quoted reply is deleted when deleted by user
* Thread reply appears in message list when `Also send to channel` box is ticked
* Thread reply is deleted when removed by participant
* Thread reply is deleted when removed by user
* Thread reply is deleted when removed by user in the thread

### 📝 Summary

As part of our Release QA we go through a bunch of manual UI scenarios that can be easily automated and catch issues early in the software development process. This work improves to build up the UI test coverage for our Chat iOS SDK.

### ☑️ Contributor Checklist
- [ ] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [ ] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [ ] This change follows zero ⚠️ policy (required)
- [ ] Comparison screenshots added for visual changes
- [ ] Affected documentation updated (docusaurus, tutorial, CMS)

### 🎁 Meme

gif

### CODE REVIEW

**Review by adolfogarza — APPROVED** [review]

File: StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift [hunk 1]
```diff
@@ -59,10 +59,10 @@ final class ParticipantRobot: Robot {
         return self
     }
     
-    // Think twice before using it
+    // Sleep in milliseconds
     @discardableResult
-    func chill(duration: UInt32) -> Self {
-        sleep(duration)
+    func chill(duration: useconds_t) -> Self {
+        usleep(duration)
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift:65]: Isn't `wait` the most common pattern here? `chill` can mean many things 
```suggestion
    func wait(duration: useconds_t) -> Self {
        usleep(duration)
```

File: StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift [hunk 2]
```diff
@@ -88,6 +88,21 @@ final class ParticipantRobot: Robot {
...
+    @discardableResult
+    func sendMultipleMessages(repeatingText text: String, count: Int) -> Self {
+        var texts = [String]()
+        for index in 1...count {
+            texts.append("\(text)-\(index)")
+        }
+
+        texts.forEach {
+            sendMessage($0)
+            chill(duration: 100)
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift:102]: This doesn't seem very efficient, why don't we just wait for the last message to appear? 
```suggestion
            waitForTextMessage(texts.last)
```

File: StreamChatUITestsAppUITests/Robots/UserRobot+Asserts.swift [hunk 3]
```diff
@@ -143,6 +143,20 @@ extension UserRobot {
...
+        for i in 1...4 {
+            let obtainKeyboardFocus = (i == 1) ? true : false
+            typeText("\(i)\n", obtainKeyboardFocus: obtainKeyboardFocus)
+            let updatedComposerHeight = composer.height
+            XCTAssertGreaterThan(updatedComposerHeight, composerHeight, file: file, line: line)
+            composerHeight = updatedComposerHeight
+        }
+        typeText("5\n6", obtainKeyboardFocus: false)
+        XCTAssertEqual(composerHeight, composer.height, file: file, line: line)
+    }
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Robots/UserRobot+Asserts.swift:159]: Can we pass the `lines` in the argument? Something like `assertComposerDoesNotGrow(moreThan numberOfLines:)`, right now if our UI Logic changes, we will need to change the function name, and the implementation as well

File: StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift [hunk 4]
```diff
@@ -59,10 +59,10 @@ final class ParticipantRobot: Robot {
         return self
     }
     
-    // Think twice before using it
+    // Sleep in milliseconds
     @discardableResult
-    func chill(duration: UInt32) -> Self {
-        sleep(duration)
+    func chill(duration: useconds_t) -> Self {
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift:64]: Can we please use `TimeInterval` here and then map to milliseconds? The common pattern in iOS is to use seconds (TimeInterval) as the unit of time

File: StreamChatUITestsAppUITests/Tests/ChannelList_Tests.swift [hunk 5]
```diff
@@ -43,7 +48,9 @@ final class ChannelList_Tests: StreamTestCase {
             deviceRobot.setConnectivity(to: .off)
         }
         WHEN("participant sends a new message") {
-            participantRobot.sendMessage(message).chill(duration: 2)
+            participantRobot
+                .sendMessage(message)
+                .chill(duration: 2000)
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Tests/ChannelList_Tests.swift:53]: I was worried about this one, I thought we were waiting 2000 seconds 😄

File: StreamChatUITestsAppUITests/Tests/MessageList_Tests.swift [hunk 6]
```diff
@@ -162,7 +162,7 @@ final class MessageList_Tests: StreamTestCase {
         AND("participant deletes the message: '\(message)'") {
             participantRobot
                 .waitForNewMessage(withText: message)
-                .chill(duration: 2)
+                .chill(duration: 2000)
```
Comment (nuno-vieira) [inline StreamChatUITestsAppUITests/Tests/MessageList_Tests.swift:165]: If we are waiting here for the new message come, why do we need to wait an additional time? 🤔

File: StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift (hunk 4 above)
Comment (nuno-vieira) [reply StreamChatUITestsAppUITests/Robots/ParticipantRobot.swift:64]: Another possible implementation is like this:

```swift
func wait(timeout: TimeInterval) {
        let runLoop = RunLoop.current
        runLoop.run(until: Date(timeIntervalSinceNow: timeout))
    }
```

File: StreamChatUITestsAppUITests/Robots/UserRobot+Asserts.swift [hunk 7]
```diff
@@ -143,6 +143,22 @@ extension UserRobot {
...
+        let composer = MessageListPage.Composer.inputField
+        var composerHeight = composer.height
+        for i in 1..<limit {
+            let obtainKeyboardFocus = (i == 1) ? true : false
+            typeText("\(i)\n", obtainKeyboardFocus: obtainKeyboardFocus)
+            let updatedComposerHeight = composer.height
+            XCTAssertGreaterThan(updatedComposerHeight, composerHeight, file: file, line: line)
+            composerHeight = updatedComposerHeight
+        }
+        typeText("5\n6", obtainKeyboardFocus: false)
```
Comment (bielikb) [inline StreamChatUITestsAppUITests/Robots/UserRobot+Asserts.swift:159]:
```suggestion
        typeText("\(limit)\n\(limit+1)", obtainKeyboardFocus: false)
```

**Review by alteral — APPROVED** [review]
🚢

**Review by nuno-vieira — APPROVED** [review]

**Pull request merged**